### PR TITLE
OpenROAD Unit Tests

### PR DESCRIPTION
**Pull request by kareefardi**
* Added tests for the following OpenROAD steps:
   * `OpenROAD.CheckAntennas`
   * `OpenROAD.DetailedPlacement`
   * `OpenROAD.DetailedRouting`
   * `OpenROAD.FillInsertion`
   * `OpenROAD.Floorplan`
   * `OpenROAD.GeneratePDN`
   * `OpenROAD.GlobalPlacement`
   * `OpenROAD.IOPlacement`
   * `OpenROAD.RCX`
   * `OpenROAD.TapDecapInsertion`
* Added warning when skipping `OpenROAD.IOPlacement` when a custom io config is set.
* Set `FP_PDN_CFG` inside the python interface not inside OpenROAD tcl script
* Test infrastructure: Added `--keep-tmp` for step tests to preserve the test tmp directory. 

---

Depends on: 
https://github.com/efabless/openlane2-step-unit-tests/pull/7

### CODE REVIEW

File: nix/openroad.nix [hunk 1]
```diff
@@ -30,6 +30,10 @@ in clangStdenv.mkDerivation rec {
     fetchSubmodules = true;
   };
 
+  patches = [
```
Comment (donn) [inline nix/openroad.nix:33]: If we merge #255 there would be no need for this

File: openlane/logging/logger.py [hunk 2]
```diff
@@ -126,7 +126,7 @@ def debug(msg: object, /, **kwargs):
     """
     if kwargs.get("stacklevel") is None:
         kwargs["stacklevel"] = 2
-    __openlane_logger.debug(msg, **kwargs)
+    print("[DEBUG]", msg)
```
Comment (donn) [inline openlane/logging/logger.py:129]: Why?

File: openlane/steps/openroad.py [hunk 3]
```diff
@@ -782,6 +783,15 @@ class GeneratePDN(OpenROADStep):
     def get_script_path(self):
         return os.path.join(get_script_dir(), "openroad", "pdn.tcl")
 
+    def run(self, state_in: State, **kwargs) -> Tuple[ViewsUpdate, MetricsUpdate]:
+        kwargs, env = self.extract_env(kwargs)
+        if self.config["FP_PDN_CFG"] is None:
+            env["FP_PDN_CFG"] = os.path.join(
+                get_script_dir(), "openroad/common/pdn_cfg.tcl"
```
Comment (donn) [inline openlane/steps/openroad.py:790]: If we're using os.path.join, may as well use three strings: "openroad" "common" "pdn_cfg.tcl"

File: openlane/steps/step.py [hunk 4]
```diff
@@ -721,7 +722,7 @@ def filename_with_counter():
         dumpable_config = copy_recursive(self.config, translator=visitor)
         dumpable_config["meta"] = {
             "openlane_version": __version__,
-            "step": self.__class__.id,
+            "step": self.__class__.id.rstrip("-0123456789"),
```
Comment (donn) [inline openlane/steps/step.py:725]: If we merge #255 we could just use `_original_id`

File: test/conftest.py [hunk 5]
```diff
@@ -125,11 +126,19 @@ def __exit__(self, exc_type, exc_value, traceback):
...
-        yield
+    if not keep_tmp:
+        with tempfile.TemporaryDirectory() as dir, chdir(dir):
+            yield
+    else:
+        dir = tempfile.mkdtemp()
+        with chdir(dir):
+            print(f"TMP: {dir}")
+            yield
+            print(f"TMP: {dir}")
```
Comment (donn) [inline test/conftest.py:141]: Why twice?

Comment (donn) [reply test/conftest.py:141]: Also, debug vestige?

File: openlane/steps/step.py (hunk 4 above)
Comment (kareefardi) [reply openlane/steps/step.py:725]: _original_id is only returned by method `with_id`. How should I utilize that here ?

Comment (donn) [reply openlane/steps/step.py:729]: i guess in `__init_subclass__` we can check if the class hasattr `_original_id` and if not, we just set it to the id?

Comment (kareefardi) [reply openlane/steps/step.py:729]: The problem is self.__class__.id varies when in a flow there a step is repeated. I am not sure where this happens but perhaps at that point `_original_id` could be set to id without numbers

Comment (donn) [reply openlane/steps/step.py]: I… implemented something

**Review by donn — APPROVED** [review]

**Pull request merged**